### ISSUE TITLE
ATTiny85 support

### ISSUE THREAD

**Issue by mish0petr0v** [RobTillaart/TinyKT0803#2]
Hello and thank you for your library.
Could you please port it to support ATTiny85 using TinyWireM.h lib for I2C?
Thanks in advance.

**Comment by RobTillaart** [comment]
Thanks for your request.

I am not familiar with the TinyWireM library and have almost no hands on experience with tiny, it is hard to say how much time it will take. Furthermore I am quite busy at the moment so have very little time to dive into this on short term.

Do you have a link to the TinyWireM library you use?

**Comment by mish0petr0v** [comment]
Thank you for your quick reply.
Here is the library in question:
`https://github.com/adafruit/TinyWireM/releases`

If you can find a spare time, I and probably many other users will be more than grateful for your efforts.

**Comment by RobTillaart** [comment]
how should it be called?
TinyKT0803 or KT0803_tiny or ???

**Comment by RobTillaart** [comment]
TinyKT0803 it will be

**Comment by RobTillaart** [comment]
@mish0petr0v 
Spend my lunch to build an initial version, 
- got some warnings 
- Serial object does not exist
- but the minimal example compiles for the ATtiny85

Please test if it works - https://github.com/RobTillaart/TinyKT0803/tree/develop

**Comment by mish0petr0v** [comment]
Thanks. I'll try it this evening.
I suspect you got "Serial object does not exist" because tis ship does not have a hardware serial port, but it can easily use the SoftwareSerial library for serial emulation.

**Comment by RobTillaart** [comment]
Yep, saw that later but no time to add

**Comment by mish0petr0v** [comment]
I have tried several ways, but every time the radio did not received the signal from transmitter.
Finally I have used a small 0,96" I2C OLED display to get a frequency feedback, using getFrequency() function..
First I tried setting frequency to 97.00 MHz and the next time I have commented the frequency row out and tried the setChannel function (channel 1940). Both ways returned one and the same number of 3276.75 which is unknown and unexpected.
If I use Nano instead of ATTiny85, the number is 35.70 for the same frequency and channel, but the radio receives a clear signal.

**Comment by RobTillaart** [comment]
Does not sound well

In the tinyWireM documentation it is explicitly saud that you need pull up resistors on the SDAand SCL.
Do you have those?

**Comment by RobTillaart** [comment]
Can you check what the ```isCinnected()``` returns?

**Comment by RobTillaart** [comment]
If it returns false you should patch the ``isConnected()``` code to see what error is returned by ```endTransmission()```

**Comment by mish0petr0v** [comment]
> Does not sound well
> 
> In the tinyWireM documentation it is explicitly saud that you need pull up resistors on the SDAand SCL. Do you have those?

I have never had a need to use them with ATTiny85 in my devices, using TinyWireM library. It should work fine without them. All my other devices do.

**Comment by RobTillaart** [comment]
Ok, fair enough

**Comment by mish0petr0v** [comment]
> If it returns false you should patch the ```isConnected()``` code to see what error is returned by ```endTransmission()```

Did not quite understood what you are asking me to do, i.e. this kind of debug is like "uncharted waters" for me, but I'll try my best.

**Comment by RobTillaart** [comment]
In the cpp file is the implementation of isConnected()

It now checks if endTransmission() is zero.
Instead you should return just the value of endTransmission(). 

This means the return type should change from void to int.
This return type change must also be done in the .h
 file.

**Comment by mish0petr0v** [comment]
OK.
Weird thing. I have added the usual 10K pull-up resistor between VDC and RESET pin and the usual 100nF capacitor between the RESET and the GND, like I doin all my ATTiny85 circuits. That changed almost nothing.
Then I tried a reset "on fly" (RESET ot GND).
After the reset, the transmitter is working, but only until next power cycle, when I have to go thrugh the reset procedure again to get it working.
p.s. The display shows the same number for getFrequency though.

**Comment by mish0petr0v** [comment]
> In the cpp file is the implementation of isConnected()
> 
> It now checks if endTransmission() is zero. Instead you should return just the value of endTransmission().
> 
> This means the return type should change from void to int. This return type change must also be done in the .h file.

Saw that now, but I'll try it tomorrow and will get back to you.

**Comment by RobTillaart** [comment]
> After the reset, the transmitter is working, but only until next power cycle,

Ok, that is very promising so we are in the right direction.

Maybe the kt0803 needs some time to do internal things before it can be started (should be in datasheet).
That could mean that a delay(xxx) in setup is needed before communicating with the device is possible.

**Comment by mish0petr0v** [comment]
I have tried everything: delays, transmission begin/end or both, before or after the setFrequency - resultless.
The ATTiny85 keeps "asking" for reboot to start the transmitter.

**Comment by RobTillaart** [comment]
So if you reboot the ATTiny85 the KT0803 starts to work.

(thinking out loud)
From the KT point of view all commands are send twice before it starts working

**Comment by mish0petr0v** [comment]
"all commands are send twice before it starts working"

Tried that out in setup section - did not worked.

**Comment by RobTillaart** [comment]
> the number is 35.70 for the same frequency and channel,

looks like it is missing a factor 2 at least =>71.4 MHz is within range.
Looking into that.

**Comment by RobTillaart** [comment]
Yep, **getChannel()** is missing a ```  channel <<= 1;```
will push an update now.

**Comment by RobTillaart** [comment]
Can you post your test sketch so I can include it as example?
(with the display)

**Comment by mish0petr0v** [comment]
It consumes way too much memory because of the String in it. I will work on the code optimization and tweaks, but I can post it anyway.

**Comment by RobTillaart** [comment]
Maybe a SWSerial example is easier?

**Comment by mish0petr0v** [comment]
```
//   https://github.com/RobTillaart/TinyKT0803/

#include <TinyKT0803.h>
#include <Tiny4kOLED.h>

TinyKT0803 FM_SEND;

void oledInit() {
  oled.begin(128, 64, sizeof(tiny4koled_init_128x64br), tiny4koled_init_128x64br);
  oled.clear();
  oled.on();
}

void oledFreq() {
  oled.clear();
  oled.setFont(FONT8X16P);
  oled.setCursor(20, 2);
  oled.print(FM_SEND.getFrequency());
}

void setup() {
  TinyWireM.begin();
  delay(50);

  FM_SEND.begin();
  FM_SEND.setFrequency(97.00);
  FM_SEND.setMute(false);

  oledInit();
  oledFreq();
}

void loop() {
  //
}
```

**Comment by RobTillaart** [comment]
Created [issue in KT0803 ](https://github.com/RobTillaart/KT0803/issues/5) for getChannel too

**Comment by mish0petr0v** [comment]
> > That could mean that a delay(xxx) in setup is needed before communicating with the device is possible.

You told me, by I was not listening. ;)
A minimum delay ot 500 millisecondsis required for the I2C bus to be initialized. The ATTiny85 works on it's minimal internal clock speed of 1 MHz (for power-saving battery-powered application purposes) and everything (I think) happens "Not so fast as we would desire to". Nano is working on 16 MHz and it does no require such delay.

```
void setup() {
  TinyWireM.begin();
  delay(500);
  
  FM_SEND.begin();
  FM_SEND.setFrequency(97.00);
  FM_SEND.setMute(false);
}
```
p.s. I will try to reduce this time delay by setting 8MHz clock speed.
p.p.s. Yep. On 8 MHz clock speed the delay could belowered to 350 milliseconds, but it is better to keep it a bit higher i.e.500-1000 just to be sure.

**Comment by RobTillaart** [comment]
> I will try to reduce this time delay by setting 8MHz clock speed.

No let that delay - I need to document this in the example and the readme.md

**Comment by mish0petr0v** [comment]
I'll try a "dynamic" frequency switch in the loop, to ensure there are no more surprises.

**Comment by mish0petr0v** [comment]
OK, it works flawlessly:

```cpp
//   https://github.com/RobTillaart/TinyKT0803/issues/2


#include <TinyKT0803.h>

TinyKT0803 FM_SEND;

void setup() {
  TinyWireM.begin();
  delay(500);
  
  FM_SEND.begin();
  FM_SEND.setMute(false);
}

void loop() {
  FM_SEND.setFrequency(93.00);
  delay(20000);
  FM_SEND.setFrequency(97.00);
  delay(20000);
  FM_SEND.setFrequency(105.00);
  delay(20000);
}
```

A few notes:
- ATTiny85 should use at least 500 ms delay to communicate properly.
- ATTiny85 should have a decoupling capacitor between RESET and GND and a 10K resistor between VDC and RESET
- ATTiny85 works on both internal clock speeds using the same time delay at comm setup.
- KT0803 is a bit cranky and the input volume from the audio device's headphones output should be lowered to 50-60%. I have somehow managed to raise this threshold on Uno/Nano using setPGA function.
- all my experiments are done with the (officially) unsupported KT0803L. I can confirm that everything works with both libraries

**Comment by RobTillaart** [comment]
Does the new **getChannel() / getFrequency()** work?

**Comment by mish0petr0v** [comment]
Never checked.
Will do now.

**Comment by RobTillaart** [comment]
Instead of the delay(500) we can use this as it adapts to clock of 1 MHz or 8 MHz

```cpp
  TinyWireM.begin();
  while (! FM_SEND.isConnected()) delay(10);   //  give I2C time to initialize, 
```

**Comment by mish0petr0v** [comment]
FM_SEND.getFrequency() shows the same numner (for 97.00 MHz): 3276.70

**Comment by RobTillaart** [comment]
mmm that needs some attention,
Does it correctly send at 97 MHz?

**Comment by mish0petr0v** [comment]
Yep. Confirmed with my phone FM receiver.
Interesting thingy: if I use getFrequency in oled.print(String(FM_SEND.getFrequency())); , the compiled sketch has 7054 bytes of size, but if i replace it with getChannel, compilation results in 5174 bytes.

p.s. The display shows 65534 on getChannel result.

**Comment by RobTillaart** [comment]
> Yep. Confirmed with my phone FM receiver. Interesting thingy: if I use getFrequency in oled.print(String(FM_SEND.getFrequency())); , the compiled sketch has 7054 bytes of size, but if i replace it with getChannel, compilation results in 5174 bytes.
> 
> p.s. The display shows 65534 on getChannel result.

Size might change because of using floating point math + String library in the conversion to frequency.

Does this result in smaller size (Then the String lib is responsible)
```cpp
oled.print(round(FM_SEND.getFrequency() * 10));
```

**Comment by mish0petr0v** [comment]
6442 bytes

**Comment by mish0petr0v** [comment]
It gets even less if I use the set frequency as string to display.

**Comment by mish0petr0v** [comment]
`oled.print(round(FM_SEND.getFrequency() * 10));`
causes 32767 result on display and breaks transmission. I do not have idea why.

**Comment by mish0petr0v** [comment]
`aaa = String(FM_SEND.setFrequency(97.00));`
and
`oled.print(aaa);`

results in 5460 bytes of size.

**Comment by RobTillaart** [comment]
> `oled.print(round(FM_SEND.getFrequency() * 10));` causes 32767 result on display and breaks transmission. I do not have idea why.

32767 is maxInt so there is overflow.  Problem is in getChannel as getFrequency is just a wrapper.

OK, I will investigate getChannel once more.

Checked the KT0803L datasheet and I see the channel bits are in the same location as the base KT0803 (K & M)

**Comment by mish0petr0v** [comment]
`while (! FM_SEND.isConnected()) delay(100);   //  give I2C time to initialize,`

doesn't work
Rolling back to

`delay(500);`

**Comment by mish0petr0v** [comment]
It turned out this oled library can display floats without additional conversion

```
float aaa = 97.00;
oled.print(aaa);
FM_SEND.setFrequency(aaa);
```
Sketch is 5426 bytes now. It is much better.

**Comment by RobTillaart** [comment]
Hypothesis:
Maybe the I2C calls need to be separated more in time (so the internal buffer is sent before it is overwritten or so).
Adding delay's between calls might improve **getChannel()** result - I found no bug yet.

**Comment by mish0petr0v** [comment]
I'll try...

Edit: tried with the followng code:

```
FM_SEND.begin();
  delay(500);
  FM_SEND.setFrequency(setfreq);
  delay(500);
  FM_SEND.setMute(false);
  delay(500);
  
  getfr = FM_SEND.getFrequency();
  delay(500);
```

The result: 3276

**Comment by RobTillaart** [comment]
> The result: 3276

That is the same 32767 in disguise...

**Comment by RobTillaart** [comment]
if you change -1 to 0, does getChannel return 0?
```cpp
int TinyKT0803::readData(uint8_t reg)
{
  TinyWireM.beginTransmission(_address);
  TinyWireM.write(reg);
  TinyWireM.endTransmission();

  if (TinyWireM.requestFrom(_address, (uint8_t) 1) == 1)
  {
    uint8_t data = TinyWireM.read();
    return data;
  }
  return 0;  //  <<<<<<<<<<<<<<<<<<<<<<<<<<<<<
}
```

**Comment by mish0petr0v** [comment]
Confirmed.

edit: should I leave it to 0 or return it to -1?

**Comment by RobTillaart** [comment]
think I found the "bug", requestFrom differs from the TwoWire behavior.

```cpp
  /*!
   * @brief Setup for receiving from secondary
   * @param slaveAddr Secondary device address
   * @param numBytes How many bytes to request
   * @return Returns 0, if there was an error, returns the error code    <<<<<<<<<<
   */
  uint8_t requestFrom(uint8_t, uint8_t);
```

**Comment by RobTillaart** [comment]
please try
```cpp
int TinyKT0803::readData(uint8_t reg)
{
  TinyWireM.beginTransmission(_address);
  TinyWireM.write(reg);
  TinyWireM.endTransmission();

  if (TinyWireM.requestFrom(_address, (uint8_t) 1) == 0)  //  <<<<<<<<<<<<  changed to 0
  {
    uint8_t data = TinyWireM.read();
    return data;
  }
  return -1;
}
```

**Comment by mish0petr0v** [comment]
Result, using getFrequency: 71

**Comment by RobTillaart** [comment]
> Result, using getFrequency: 71

And what was the frequency set?

**Comment by RobTillaart** [comment]
pushed a new version of the develop branch including the requestFrom() fix.

**Comment by RobTillaart** [comment]
(I will be offline for ~3hrs)

**Comment by mish0petr0v** [comment]
97 MHz.

I have noticed the following in KT0803L pdf and I do not know if itmakesa difference:

```
Please note that the default channel of KT0803L is 86MHz instead of 89.7MHz in KT0803K and
KT0803M. 
```

**Comment by RobTillaart** [comment]
As we set freq before fetching it the default is less relevant.

97 and 71 are still too far apart.

Can you test setChannel. GetChannel  if it reads back same?

**Comment by mish0petr0v** [comment]
Using setChannel and getChannel for channel 1940 (97.00 MHz): 1428

**Comment by RobTillaart** [comment]
1940 = 111 1001 0100
1428 = 101 1001 0100

So one bit got lost during getChannel().

Test if we always lose the same bit:

What does getCHannel() return when you use setChannel(2000) ?

**Comment by RobTillaart** [comment]
Added extra functions in the KT0803K class

**Comment by mish0petr0v** [comment]
>What does getCHannel() return when you use setChannel(2000) ?

4046

**Comment by RobTillaart** [comment]
2000 = 0111 1101 0000
4046 = 1010 0011 1010

makes no sense at all, not even partially look alike (time to think really deep)

**Comment by mish0petr0v** [comment]
`ATTiny85 needs a decoupling capacitor between RESET and GND...`

Forgot to tell you the cap value - about 100-300 nF.

**Comment by mish0petr0v** [comment]
> 2000 = 0111 1101 0000 4046 = 1010 0011 1010
> 
> makes no sense at all, not even partially look alike (time to think really deep)

Are you sure you read the entire register and the proper start-end addresses?

**Comment by RobTillaart** [comment]
> Are you sure you read the entire register and the proper start-end addresses?

It is even worse, the channel is divided over three registers
Writing works fine, reading it back is - lets say time consuming to get it right :)

![image](https://github.com/RobTillaart/TinyKT0803/assets/462844/47d65fac-f7f9-491e-acfe-4d7deb510598)

I'm gonna write a register dumper, to be able to analyze this device.

**Comment by RobTillaart** [comment]
Written a sort of register dumper with 3 seconds in between using the oled code.

Can you give it a try?

**Comment by mish0petr0v** [comment]
> Written a sort of register dumper with 3 seconds in between using the oled code.
> 
> Can you give it a try?

232
232
232

**Comment by RobTillaart** [comment]
```
 232 =   11101000
2000 = 11111010000
       xx++++++++x
```
So register 0 is correct and it seems to always read register 0.

Dive into documentation...

**Comment by RobTillaart** [comment]
Can you change the readData function to

```cpp
int TinyKT0803::readData(uint8_t reg)
{
  TinyWireM.beginTransmission(_address);
  TinyWireM.write(reg);
  TinyWireM.endTransmission(false);  //   <<<<<<<<<<<<<<<<<< explicit no STOP

  //  Behaviour differs from TwoWire class
  int rv = TinyWireM.requestFrom(_address, (uint8_t) 1);
  if (rv == 0)
  {
    uint8_t data = TinyWireM.read();
    return data;
  }
  return rv;
}
```

and check getChannel() again? 

hypothesis is that the STOP signal resets the register index to zero.

Note: The STOP bit is implicit in endTransmission()

**Comment by mish0petr0v** [comment]
232
195
64

**Comment by RobTillaart** [comment]
Yes, that looks good
```
 232 =   11101000
2000 = 11111010000
       xx++++++++x = 8 middle bits from 2000

195 = 11000011
           +++ = msb 3 bits from 2000

 64 = 01000000
      + = lsb 1 bit from 2000
```
think we nailed it.

now getFrequency should work too, can you test?

**Comment by mish0petr0v** [comment]
> think we nailed it.

Not "we", but You. ;)

> now getFrequency should work too, can you test?

Yep. both show back correctly: 1940 and 97
(The getFrequency returns int instead of float though.)

p.s.Oops, sorry. My bad. :)
Forgot to add float instead of long to my screen function
`void screen(float abc, int ax, int ay, int font)`

**Comment by RobTillaart** [comment]
You did a lot of testing, without that it would still not work.

Tomorrow I will cleanup the code and release and register the library. I already patched the kt0803 library with same fix.

**Comment by RobTillaart** [comment]
You should now be able to set a frequency of e.g 97.15. (using the kt0803k object)

Would it make sense for you to have a derived kt0803L class?

**Comment by mish0petr0v** [comment]
`You should now be able to set a frequency of e.g 97.15. (using the kt0803k object)`
On my other sketch, using Arduino Nano I've set a "fine tune" function, that sets the offset in transmision i.e. offset of 10 means for 97 MHz to set between 96.90 and 97.10. It happens that some radios receive better on 50 kHz lower setting. But What happens actually when I set the fine tune to 96.97 MHz? Does it really transmits to that frequency or the library is rounding it to the closest possible one? Because the resulting channel becomes a fractional number.

`Would it make sense for you to have a derived kt0803L class?`
I cannot quite understand the question? What do you meanby "derived"?

**Comment by mish0petr0v** [comment]
Channel 1939.2 is rounding to 1938.
Odd...

**Comment by mish0petr0v** [comment]
Channel 1939 (96.95 MHz) rounds to 1938 (96.90 MHz).
Really odd...

**Comment by RobTillaart** [comment]
In the sketch you have the line
TinyKT0803 FM_SEND;

that creates the object

Change that to
TinyKT0803K FM_SEND;

And you can set the frequency in steps of 50 KHz.
So 96.00, 96.05, 96.10, 96.15 etc.

**Comment by mish0petr0v** [comment]
> In the sketch you have the line TinyKT0803 FM_SEND;
> 
> that creates the object
> 
> Change that to TinyKT0803K FM_SEND;
> 
> And you can set the frequency in steps of 50 KHz. So 96.00, 96.05, 96.10, 96.15 etc.

Does not work. Channel 1939 cannot be set and it "rounds down" to 1938.

**Comment by mish0petr0v** [comment]
Is it on purpose you are missing getFrequency class from 0803K section?

I've added it to fit my own needs (less vars and code lines), but I think it could be better if you add it to your release.
And maybe it would bebnetter you to remove the register_dumper sketch, it was only for debuging purposes. ;)

**Comment by RobTillaart** [comment]
Good morning,
The KT0803 device can only set even channels as it does have the CHSEL[0] bit in register2 always set to zero. Even if you write a 1 to it.
In frequency terms this means the resolution is 100 KHz or 0.1MHz. This results in rounding down.

The KT0803K device (and L and M models) can set the CHSEL[0] bit so they can do the odd channels too. There is no rounding of channels.
In frequency terms this is a resolution of 50KHz or 0.05 MHz.

**Comment by RobTillaart** [comment]
Yes, the debugger is one of the things to reconsider. Sure I will keep a copy but agree it has little value for most users unless they want to extend the classes.

**Comment by RobTillaart** [comment]
> Is it on purpose you are missing getFrequency class from 0803K section?

The KT0803K class inherits this function (and many others) from KT0803.  
The KT0803K has some specific function like **setMono()**.

**Comment by RobTillaart** [comment]
@mish0petr0v 

Released the 0.1.0 version, so I close this issue.
If new problems pop up, please open a new (specific) issue.

**Comment by RobTillaart** [comment]
Published 0.1.0 release on platformIO and Arduino Library manager.
